$NetBSD$

Substitute whoami for id across SunOS platforms.
Silence sudo (unneeded screen waste).
Make sure we have enough file descriptors (on SunOS).
Warn against using riak to start/stop rather than SMF.
--- rel/files/riak.orig	2012-08-06 22:02:45.000000000 +0000
+++ rel/files/riak
@@ -13,6 +13,12 @@ if [ `uname -s` = 'SunOS' -a "${POSIX_SH
 fi
 unset POSIX_SHELL # clear it so if we invoke other scripts, they run as ksh as well
 
+if [ `uname -s` = 'SunOS' ]; then
+    WHOAMI_PROG="/usr/xpg4/bin/id -un"
+else
+    WHOAMI_PROG="whoami"
+fi    
+
 RUNNER_SCRIPT_DIR={{runner_script_dir}}
 RUNNER_SCRIPT=${0##*/}
 
@@ -27,7 +33,7 @@ PLATFORM_DATA_DIR={{platform_data_dir}}
 SSL_DIST_CONFIG=$PLATFORM_DATA_DIR/ssl_distribution.args_file
 RIAK_VERSION="git"
 
-WHOAMI=$(whoami)
+WHOAMI=$(${WHOAMI_PROG})
 
 # Make sure this script is running as the appropriate user
 if ([ "$RUNNER_USER" ] && [ "x$WHOAMI" != "x$RUNNER_USER" ]); then
@@ -36,8 +42,18 @@ if ([ "$RUNNER_USER" ] && [ "x$WHOAMI" !
         echo "sudo doesn't appear to be installed and your EUID isn't $RUNNER_USER" 1>&2
         exit 1
     fi
-    echo "Attempting to restart script through sudo -H -u $RUNNER_USER"
-    exec sudo -H -u $RUNNER_USER -i $RUNNER_SCRIPT_DIR/$RUNNER_SCRIPT $@
+    exec sudo -H -u $RUNNER_USER $RUNNER_SCRIPT_DIR/$RUNNER_SCRIPT $@
+fi
+
+# Make sure we preload extendedFILE.so.1 so that 32bit eleveldb works on SunOS
+export LD_PRELOAD_32=${LD_PRELOAD_32}:/usr/lib/extendedFILE.so.1
+
+# Make sure we have access to enough file descriptors
+ULIMIT_S=$(prctl -n process.max-file-descriptor -t basic -P $$ | awk '/max-file-descriptor/ { print $3 }')
+ULIMIT_H=$(prctl -n process.max-file-descriptor -t priv -P $$ | awk '/max-file-descriptor/ { print $3 }')
+if [ ${ULIMIT_S} -lt ${ULIMIT_H} ]; then
+    echo "Trying to raise the file descriptor limit to maximum allowed."
+    prctl -n process.max-file-descriptor -t basic -v ${ULIMIT_H} $$ || true
 fi
 
 # Warn the user if ulimit -n is less than 1024
@@ -102,6 +118,14 @@ rm -f $SSL_DIST_CONFIG
 sed -n '/Begin SSL distribution items/,/End SSL distribution items/p' \
     $RUNNER_ETC_DIR/vm.args > $SSL_DIST_CONFIG
 
+case "$1" in
+    start|stop|restart|reboot)
+        echo "***"
+        echo "Warning: please use 'svcadm disable riak' instead"
+        echo "***"
+    ;;
+esac
+
 # Check the first argument for instructions
 case "$1" in
     start)
